File: src/main/java/com/smarthome/backend/server/actions/ActionRunnable.java
package com.smarthome.backend.server.actions;

import java.util.function.Consumer;

/**
 * Runnable-Implementierung für die Ausführung von Actions.
 * 
 * Diese Klasse kapselt eine Action und ermöglicht deren Ausführung
 * als Runnable, z.B. für Threads, ExecutorServices oder Scheduler.
 */
public abstract class ActionRunnable {

    protected final Runnable listener;
    protected final Consumer<Object> listenerWithParam;

    public ActionRunnable(Runnable listener) {
        this.listener = listener;
        this.listenerWithParam = null;
    }

    public ActionRunnable(Consumer<Object> listenerWithParam) {
        this.listener = null;
        this.listenerWithParam = listenerWithParam;
    }

    public void run() {
        if (listener != null) {
            listener.run();
        } else if (listenerWithParam != null) {
            listenerWithParam.accept(null);
        }
    }

    public void run(Object value) {
        if (listenerWithParam != null) {
            listenerWithParam.accept(value);
        } else if (listener != null) {
            listener.run();
        }
    }

}

File: src/main/node/com/smarthome/backend/server/actions/actionRunnable.ts
export class ActionRunnable {
  protected listener?: () => void;
  protected listenerWithParam?: (value: unknown) => void;

  constructor(listener: () => void);
  constructor(listenerWithParam: (value: unknown) => void);
  constructor(callback: (() => void) | ((value: unknown) => void)) {
    if (callback.length === 0) {
      this.listener = callback as () => void;
    } else {
      this.listenerWithParam = callback as (value: unknown) => void;
    }
  }

  run() {
    if (this.listener) {
      this.listener();
    } else if (this.listenerWithParam) {
      this.listenerWithParam(null);
    }
  }

  runWithValue(value: unknown) {
    if (this.listenerWithParam) {
      this.listenerWithParam(value);
    } else if (this.listener) {
      this.listener();
    }
  }
}

